### PR TITLE
gh-106368: Argument clinic: add tests for more failure paths

### PR DESCRIPTION
**Pull request by AlexWaygood**
And fix a minor bug I discovered along the way.

This gets us up to 95% coverage.

* Issue: gh-106368

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -5194,13 +5194,14 @@ def bad_node(self, node: ast.AST) -> None:
...
-                        if value is unspecified:
-                            fail("'unspecified' is not a legal default value!")
                     except NameError:
                         pass # probably a named constant
                     except Exception as e:
                         fail("Malformed expression given as default value "
                              f"{default!r} caused {e!r}")
+                    else:
+                        if value is unspecified:
+                            fail("'unspecified' is not a legal default value!")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:5217]: Without this change, the default value in

```
            module os
            os.access
                follow_symlinks: int(c_default='MAXSIZE') = unspecified
```

was being reported as a "malformed expression" in the error message, which isn't really true. (the `try`/`except` above was catching the `ClinicError` raised by the `fail()` call in the `if value is unspecified` branch)

File: Lib/test/test_clinic.py [hunk 2]
```diff
@@ -1152,6 +1244,20 @@ def test_left_group(self):
                 Attributes for the character.
         """)
 
+    def test_unmatched_right_square_paren(self):
+        err = "Function 'addch' has a ']' without a matching '['."
```
Comment (erlend-aasland) [inline Lib/test/test_clinic.py]: I thought we had a test for this already?

File: Tools/clinic/clinic.py (hunk 1 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:5217]: Good catch!

**Review by erlend-aasland — APPROVED** [review]
Looks good. BTW, can you remove `test_parameters_required_after_depr_star` (line 1471)? I introduced an equal named, but slightly better test in #95151). Makes sense to bake it into this PR, rather than creating a new PR specifically for that removal.

File: Lib/test/test_clinic.py (hunk 2 above)
Comment (AlexWaygood) [reply Lib/test/test_clinic.py]: You're right. Not sure what went on there. Maybe a merge race, maybe a lack of caffeine :)

I'll remove this test.

**Comment by AlexWaygood** [comment]
> Looks good. BTW, can you remove `test_parameters_required_after_depr_star` (line 1471)? I introduced an equal named, but slightly better test in #95151). Makes sense to bake it into this PR, rather than creating a new PR specifically for that removal.

Hmm, the new test that you added fails if I remove `test_parameters_required_after_depr_star` on line 1471. The new test method that you added is currently never run, because it is immediately overridden by the old test method later on in the class definition before unittest even tries to run any of the tests.

We really need to start running pyflakes on `Tools/clinic/` and `test_clinic.py` in CI :)

**Comment by erlend-aasland** [comment]
Well, remove the new test, then 😄

**Review by erlend-aasland — APPROVED** [review]

**Pull request merged**